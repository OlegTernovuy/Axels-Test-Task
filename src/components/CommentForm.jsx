import { useState } from 'react';
import { Button, Rating, TextField } from '@mui/material';

import { BoxForm } from '../styled/StyledCommentForm';

const CommentForm = () => {
  const [ratingValue, setRatingValue] = useState(0);

  return (
    <BoxForm component="form">
      <TextField label="comment" variant="outlined" />
      <Rating
        name="simple-controlled"
        value={ratingValue}
        onChange={(event, newValue) => {
          setRatingValue(newValue);
        }}
      />
      <Button variant="outlined" type="submit">
        Submit
      </Button>
    </BoxForm>
  );
};

export default CommentForm;
