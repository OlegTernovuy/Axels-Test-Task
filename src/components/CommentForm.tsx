import { useFormik } from 'formik';
import * as yup from 'yup';
import { Button, Rating, TextField } from '@mui/material';

import { BoxForm } from '../styled/StyledCommentForm';

interface FormValues {
  comment: string;
  rating: number;
}

const validationSchema = yup.object({
  comment: yup
    .string()
    .min(5, 'Сomment must contain at least 5 char')
    .max(25, 'Сomment should not contain more than 25 char')
    .required('Comment is required'),
  rating: yup.number().min(1, 'Rating is required'),
});

const CommentForm = () => {
  const formik = useFormik({
    initialValues: {
      comment: '',
      rating: 0,
    } as FormValues,
    validationSchema: validationSchema,
    onSubmit: (values: FormValues, { resetForm }) => {
      resetForm();
      console.log(values);
    },
  });

  return (
    <BoxForm onSubmit={formik.handleSubmit}>
      <TextField
        label='comment'
        name='comment'
        variant='outlined'
        value={formik.values.comment}
        onChange={formik.handleChange}
        error={formik.touched.comment && Boolean(formik.errors.comment)}
        helperText={formik.touched.comment && formik.errors.comment}
      />
      <Rating
        name='rating'
        value={Number(formik.values.rating)}
        onChange={formik.handleChange}
      />
      <span>
        {formik.touched.rating && Boolean(formik.errors.rating)
          ? formik.errors.rating
          : null}
      </span>
      <Button variant='outlined' fullWidth type='submit'>
        Submit
      </Button>
    </BoxForm>
  );
};

export default CommentForm;
