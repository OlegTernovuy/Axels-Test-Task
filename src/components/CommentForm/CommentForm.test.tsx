import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';

import CommentForm from './CommentForm';

describe('Comment form snapshot component', () => {
    test('renders CommentForm correctly', () => {
        render(<CommentForm />);
        expect(screen.getByLabelText(/comment/i)).toBeInTheDocument();
        expect(
            screen.getByRole('radio', { name: /Empty/i })
        ).toBeInTheDocument();
        expect(
            screen.getByRole('button', { name: /submit/i })
        ).toBeInTheDocument();
    });

    test('shows validation errors on submit with empty fields', async () => {
        render(<CommentForm />);

        fireEvent.click(screen.getByRole('button', { name: /submit/i }));

        expect(
            await screen.findByText(/comment is required/i)
        ).toBeInTheDocument();
        expect(
            await screen.findByText(/rating is required/i)
        ).toBeInTheDocument();
    });

    test('submits form with correct values', () => {
        render(<CommentForm />);

        fireEvent.change(screen.getByLabelText(/comment/i), {
            target: { value: 'This is a test comment' },
        });
        fireEvent.change(screen.getByRole('radio', { name: /5 Stars/i }), {
            target: { value: 5 },
        });

        fireEvent.click(screen.getByRole('button', { name: /submit/i }));

        expect(
            screen.queryByText(/comment is required/i)
        ).not.toBeInTheDocument();
        expect(
            screen.queryByText(/rating is required/i)
        ).not.toBeInTheDocument();
    });

    test('shows validation error for too short comment', async () => {
        render(<CommentForm />);

        fireEvent.change(screen.getByLabelText(/comment/i), {
            target: { value: '1234' },
        });

        fireEvent.click(screen.getByRole('button', { name: /submit/i }));

        expect(
            await screen.findByText(/Ð¡omment must contain at least 5 char/i)
        ).toBeInTheDocument();
    });

    it('Matches Snapshot', () => {
        const commentFormComponent = renderer.create(<CommentForm />).toJSON();
        expect(commentFormComponent).toMatchSnapshot();
    });
});
