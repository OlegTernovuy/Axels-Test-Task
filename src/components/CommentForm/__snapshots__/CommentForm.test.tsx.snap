// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentForm component CommentForm matches Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="sc-aYaIB jNoMlc"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":rs:"
            id=":rs:-label"
          >
            comment
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":rs:"
              name="comment"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  comment
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <span
          class="MuiRating-root MuiRating-sizeMedium css-1qqgbpl-MuiRating-root"
        >
          <label
            class="css-dqr9h-MuiRating-label"
            for=":ru:"
          >
            <span
              class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarBorderIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                />
              </svg>
            </span>
            <span
              class="MuiRating-visuallyHidden"
            >
              1 Star
            </span>
          </label>
          <input
            class="MuiRating-visuallyHidden"
            id=":ru:"
            name="rating"
            type="radio"
            value="1"
          />
          <label
            class="css-dqr9h-MuiRating-label"
            for=":rv:"
          >
            <span
              class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarBorderIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                />
              </svg>
            </span>
            <span
              class="MuiRating-visuallyHidden"
            >
              2 Stars
            </span>
          </label>
          <input
            class="MuiRating-visuallyHidden"
            id=":rv:"
            name="rating"
            type="radio"
            value="2"
          />
          <label
            class="css-dqr9h-MuiRating-label"
            for=":r10:"
          >
            <span
              class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarBorderIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                />
              </svg>
            </span>
            <span
              class="MuiRating-visuallyHidden"
            >
              3 Stars
            </span>
          </label>
          <input
            class="MuiRating-visuallyHidden"
            id=":r10:"
            name="rating"
            type="radio"
            value="3"
          />
          <label
            class="css-dqr9h-MuiRating-label"
            for=":r11:"
          >
            <span
              class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarBorderIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                />
              </svg>
            </span>
            <span
              class="MuiRating-visuallyHidden"
            >
              4 Stars
            </span>
          </label>
          <input
            class="MuiRating-visuallyHidden"
            id=":r11:"
            name="rating"
            type="radio"
            value="4"
          />
          <label
            class="css-dqr9h-MuiRating-label"
            for=":r12:"
          >
            <span
              class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarBorderIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                />
              </svg>
            </span>
            <span
              class="MuiRating-visuallyHidden"
            >
              5 Stars
            </span>
          </label>
          <input
            class="MuiRating-visuallyHidden"
            id=":r12:"
            name="rating"
            type="radio"
            value="5"
          />
          <label
            class="MuiRating-label MuiRating-pristine css-dqr9h-MuiRating-label"
          >
            <input
              class="MuiRating-visuallyHidden"
              id="rating-empty"
              name="rating"
              type="radio"
              value=""
            />
            <span
              class="MuiRating-visuallyHidden"
            >
              Empty
            </span>
          </label>
        </span>
        <span />
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-xsqe0f-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Submit
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="sc-aYaIB jNoMlc"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":rs:"
          id=":rs:-label"
        >
          comment
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":rs:"
            name="comment"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                comment
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <span
        class="MuiRating-root MuiRating-sizeMedium css-1qqgbpl-MuiRating-root"
      >
        <label
          class="css-dqr9h-MuiRating-label"
          for=":ru:"
        >
          <span
            class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarBorderIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
              />
            </svg>
          </span>
          <span
            class="MuiRating-visuallyHidden"
          >
            1 Star
          </span>
        </label>
        <input
          class="MuiRating-visuallyHidden"
          id=":ru:"
          name="rating"
          type="radio"
          value="1"
        />
        <label
          class="css-dqr9h-MuiRating-label"
          for=":rv:"
        >
          <span
            class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarBorderIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
              />
            </svg>
          </span>
          <span
            class="MuiRating-visuallyHidden"
          >
            2 Stars
          </span>
        </label>
        <input
          class="MuiRating-visuallyHidden"
          id=":rv:"
          name="rating"
          type="radio"
          value="2"
        />
        <label
          class="css-dqr9h-MuiRating-label"
          for=":r10:"
        >
          <span
            class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarBorderIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
              />
            </svg>
          </span>
          <span
            class="MuiRating-visuallyHidden"
          >
            3 Stars
          </span>
        </label>
        <input
          class="MuiRating-visuallyHidden"
          id=":r10:"
          name="rating"
          type="radio"
          value="3"
        />
        <label
          class="css-dqr9h-MuiRating-label"
          for=":r11:"
        >
          <span
            class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarBorderIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
              />
            </svg>
          </span>
          <span
            class="MuiRating-visuallyHidden"
          >
            4 Stars
          </span>
        </label>
        <input
          class="MuiRating-visuallyHidden"
          id=":r11:"
          name="rating"
          type="radio"
          value="4"
        />
        <label
          class="css-dqr9h-MuiRating-label"
          for=":r12:"
        >
          <span
            class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarBorderIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
              />
            </svg>
          </span>
          <span
            class="MuiRating-visuallyHidden"
          >
            5 Stars
          </span>
        </label>
        <input
          class="MuiRating-visuallyHidden"
          id=":r12:"
          name="rating"
          type="radio"
          value="5"
        />
        <label
          class="MuiRating-label MuiRating-pristine css-dqr9h-MuiRating-label"
        >
          <input
            class="MuiRating-visuallyHidden"
            id="rating-empty"
            name="rating"
            type="radio"
            value=""
          />
          <span
            class="MuiRating-visuallyHidden"
          >
            Empty
          </span>
        </label>
      </span>
      <span />
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-xsqe0f-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Submit
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
